Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -0,0 +1,709 @@
+				var breakDistance = breakScale * k_DirectSurfaceSnapBreakDist;
+
+				if (surfaceSnapping && PerformDirectSurfaceSnapping(ref position, ref rotation, targetPosition, state, targetRotation, breakDistance))
+					return true;
+
+				if (groundSnapping && PerformGroundSnapping(ref position, ref rotation, targetPosition, targetRotation, state, breakScale * k_DirectGroundSnapMin, breakScale * k_DirectGroundSnapMax))
+					return true;
+			}
+
+			position = targetPosition;
+			rotation = targetRotation;
+
+			return false;
+		}
+
+		bool PerformManipulatorSurfaceSnapping(Transform rayOrigin, ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, SnappingState state, Quaternion targetRotation, float breakDistance)
+		{
+			SetupIgnoreList(m_IgnoreList, playerHeadObjects, state.go);
+
+			var bounds = state.identityBounds;
+			var boundsExtents = bounds.extents;
+			var projectedExtents = Vector3.Project(boundsExtents, Vector3.down);
+			var offset = projectedExtents - bounds.center;
+			var rotationOffset = Quaternion.AngleAxis(90, Vector3.right);
+			var startRotation = state.startRotation;
+			var upVector = startRotation * Vector3.back;
+			var maxRayLength = k_SurfaceSnappingMaxRayLength * this.GetViewerScale();
+
+			var ray = new Ray(rayOrigin.position, rayOrigin.forward);
+			return PerformSurfaceSnapping(ray, ref position, ref rotation, targetPosition, state, offset, targetRotation, rotationOffset , upVector, m_IgnoreList, breakDistance, maxRayLength)
+				|| TryBreakSurfaceSnapping(ref position, ref rotation, targetPosition, startRotation, state, breakDistance);
+		}
+
+		bool PerformDirectSurfaceSnapping(ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, SnappingState state, Quaternion targetRotation, float breakDistance)
+		{
+			SetupIgnoreList(m_IgnoreList, playerHeadObjects, state.go);
+			var bounds = state.identityBounds;
+			var boundsCenter = bounds.center;
+			for (int i = 0; i < 6; i++)
+			{
+				var upVector = GetUpVector(i, targetRotation);
+				var directionVector = GetDirection(i);
+				var rotationOffset = GetRotation(i);
+				var ray = new Ray(targetPosition + targetRotation * boundsCenter, targetRotation * directionVector);
+
+				var boundsExtents = bounds.extents;
+				var projectedExtents = Vector3.Project(boundsExtents, directionVector);
+				var raycastDistance = projectedExtents.magnitude * k_DirectSurfaceSearchScale;
+				var offset = -boundsCenter;
+				if (i > 2)
+					offset -= projectedExtents;
+				else
+					offset += projectedExtents;
+
+				if (PerformSurfaceSnapping(ray, ref position, ref rotation, targetPosition, state, offset, targetRotation, rotationOffset, upVector, m_IgnoreList, breakDistance, raycastDistance))
+					return true;
+			}
+
+			if (TryBreakSurfaceSnapping(ref position, ref rotation, targetPosition, targetRotation, state, breakDistance))
+				return true;
+
+			return false;
+		}
+
+		static Vector3 GetDirection(int i)
```

Comment:
Instead of these methods can you create a constant array of a struct that defines the direction, rotation, up vector, etc. and iterate over that instead of having a magic number in your for loops?